clean_ceph_vols.py

```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env python
+"""
+Utility to cleanup orphaned volumes.
+
+"""
+from gevent import monkey, sleep
+
+monkey.patch_all()
+import datetime
+import yaml
+import sys
+from docopt import docopt
+from libcloud.common.types import LibcloudError
+from libcloud.compute.providers import get_driver
+from libcloud.compute.types import Provider
+from libcloud.common.exceptions import BaseHTTPError
+
+
+from ceph.parallel import parallel
+from utility.retry import retry
+
+
+doc = """
+Utility to cleanup orphaned volumes.
+
+ Usage:
+  clean_ceph_vols.py [--osp-cred <file>]
+
+Options:
+  --osp-cred <file>                 openstack credentials as separate file [default: osp/osp-cred-ci-2.yaml]
+"""
+
+
+def get_openstack_driver(yaml):
+    OpenStack = get_driver(Provider.OPENSTACK)
+    glbs = yaml.get("globals")
+    os_cred = glbs.get("openstack-credentials")
+    username = os_cred["username"]
+    password = os_cred["password"]
+    auth_url = os_cred["auth-url"]
+    auth_version = os_cred["auth-version"]
+    tenant_name = os_cred["tenant-name"]
+    service_region = os_cred["service-region"]
+    domain_name = os_cred["domain"]
+    tenant_domain_id = os_cred["tenant-domain-id"]
+    driver = OpenStack(
+        username,
+        password,
+        ex_force_auth_url=auth_url,
+        ex_force_auth_version=auth_version,
+        ex_tenant_name=tenant_name,
+        ex_force_service_region=service_region,
+        ex_domain_name=domain_name,
+        ex_tenant_domain_id=tenant_domain_id,
+    )
+    return driver
+
+
+def volume_cleanup(volume, driver):
+    print("Removing volume", volume.name)
+    errors = {}
+    sleep(1)
+    timeout = datetime.timedelta(seconds=200)
+    starttime = datetime.datetime.now()
+    while True:
+        try:
+            volobj = driver.ex_get_volume(volume.id)
+            driver.detach_volume(volobj)
+            sleep(30)
+            driver.destroy_volume(volobj)
+            break
+        except BaseHTTPError as e:
+            print(e)
+            errors.update({volume.name: e.message})
+            if errors:
+                if datetime.datetime.now() - starttime > timeout:
+                    for vol, err in errors.items():
+                        print("Error destroying", vol, ":", err)
+                    return 1
+
+
+def cleanup_ceph_vols(osp_cred):
+    """
+    Cleanup stale volues with satus deleting, available, error, reserved
+    """
+    vol_states = ["deleting", "available", "error", "reserved"]
```

Comment:
Just verifying that we really want to delete volumes that are "available" and "reserved" here?

It sounds like we would want to check the age of the volume in that case, or this could race with other tools that expect to create and attach volumes?

Reply:
Noted. Will try to get list of Vols >n days before declaring candidates for deletion